Add isemoji function to Unicode stdlib and export it

archermarx opened the pull request:
Resolves #38063

archermarx committed:
Add isemoji function and export it

archermarx committed:
Add emoji tests

archermarx committed:
Fix tests involving unexported zero-width joiner

archermarx commented:
I'm not sure why the tests are failing here. Looking at the buildbot results, it doesn't seem to have anything to do with code I introduced? The windows versions fail on a different test, and the linux and mac versions pass all tests but error at a later point due to something involving python (i think?). Anyone know what's going on?

StefanKarpinski commented on stdlib/Unicode/src/Unicode.jl:
```diff
@@ -89,4 +89,44 @@ letter combined with an accent mark is a single grapheme.)
+    0x1F600:0x1F64F,  # Emoticons
+    0x1F300:0x1F5FF,  # Misc Symbols and Pictographs
+    0x1F680:0x1F6FF,  # Transport and Map
+    0x2600:0x26FF,    # Misc symbols
+    0x2700:0x27BF,    # Dingbats
+    0xFE00:0xFE0F,    # Variation Selectors
```
```suggestion
    0x02600:0x026FF,  # Misc symbols
    0x02700:0x027BF,  # Dingbats
    0x0FE00:0x0FE0F,  # Variation Selectors
```
Just for nicer alignment.

StefanKarpinski commented on stdlib/Unicode/src/Unicode.jl:
```diff
@@ -89,4 +89,44 @@ letter combined with an accent mark is a single grapheme.)
+    return false
+end
+
+function isemoji(s::AbstractString)
```
```suggestion
function isemoji(s::AbstractString)
    isempty(s) && return true
```
It's a bit unclear what to do if the string is empty: on the one hand, it contains no emoji; on the other hand it contains no characters that aren't emoji. Either way, the empty string needs to be checked for before you can do `s[end]`.

archermarx replied:
Thanks for the catch! This is tricky. `isascii` was my mold for this, and that returns true when an empty string is passed. However, it does feel a bit weird to return true. I think the best thing is to update the docstring to make this behavior clear

stevengj replied:
I agree that `true` is the correct answer here.

archermarx committed:
Update stdlib/Unicode/src/Unicode.jl

Co-authored-by: Stefan Karpinski <stefan@karpinski.org>

archermarx committed:
Update stdlib/Unicode/src/Unicode.jl

Handle empty strings

Co-authored-by: Stefan Karpinski <stefan@karpinski.org>

StefanKarpinski commented on stdlib/Unicode/test/runtests.jl:
```diff
@@ -404,3 +404,23 @@ end
+    @test isemoji("🚴🏿")
+    @test !isemoji("👨‍👧" * Unicode.ZWJ)
+    @test isemoji("🛌" * Unicode.ZWJ * '😎')
+    @test !isemoji("🤦🏽" * Unicode.ZWJ * Unicode.ZWJ * '😎')
```
```suggestion
    @test !isemoji("🤦🏽" * Unicode.ZWJ * Unicode.ZWJ * '😎')
    @test isemoji("")
```
May as well explicitly test for the empty string case.

stevengj commented on stdlib/Unicode/src/Unicode.jl:
```diff
@@ -89,4 +89,45 @@ letter combined with an accent mark is a single grapheme.)
+    0x02600:0x026FF,  # Misc symbols
+    0x02700:0x027BF,  # Dingbats
+    0x0FE00:0x0FE0F,  # Variation Selectors
+    0x1F900:0x1F9FF,   # Supplemental Symbols and Pictographs
```
Should this be generated from https://unicode.org/Public/emoji/13.1/emoji-sequences.txt somehow?     Or at least have a test that checks that everything from https://unicode.org/Public/emoji/13.1/emoji-sequences.txt is in this set?

archermarx replied:
That's a good idea. I'll get to work on this. Is it ok to include a local copy of this file in the /Unicode/test/ directory?

stevengj replied:
Can't we just download it during the test?

StefanKarpinski replied:
We can but it may make CI flaky. Let's try that and see how it goes. If it's flaky we may need to cache a copy but that can be dealt with separately.

archermarx replied:
I implemented y'all's suggestions, so now we download https://www.unicode.org/Public/13.0.0/ucd/emoji/emoji-data.txt from the unicode website and use that to generate a const containing emoji ranges, which are then tested against. In the tests, we download the https://unicode.org/Public/emoji/13.1/emoji-sequences.txt and https://www.unicode.org/Public/emoji/13.1/emoji-zwj-sequences.txt files, which we use to build a list of all valid emoji, which we then test the isemoji function against.

stevengj replied:
I agree that we shouldn't return `true` for currently unassigned codepoints, but we can check for that simply by returning `false` if `Unicode.category_code(char) == Unicode.UTF8PROC_CATEGORY_CN`

archermarx committed:
Update isemoji documentation

archermarx committed:
Update stdlib/Unicode/test/runtests.jl

Co-authored-by: Stefan Karpinski <stefan@karpinski.org>

archermarx committed:
Redo isemoji to fix missed corner cases

The previous range list turned out not to be exhaustive. Now, we read the single emoji ranges directly from the unicode consortium website. Additionally added support for emojis that contain variation selectors ('\uFE0F')

archermarx committed:
Add new emoji tests

archermarx committed:
Update documentation

archermarx commented:
Something's broken now and I'm not sure why.

stevengj commented on stdlib/Unicode/src/Unicode.jl:
```diff
@@ -89,4 +89,86 @@ letter combined with an accent mark is a single grapheme.)
 """
 graphemes(s::AbstractString) = Base.Unicode.GraphemeIterator{typeof(s)}(s)
 
+const emoji_data = download("https://www.unicode.org/Public/13.0.0/ucd/emoji/emoji-data.txt")
```
I don’t think we want to store this entire string in the compiled library. You should just download it when you parse the data, maybe?

archermarx replied:
Yeah, that makes sense.

stevengj replied:
Your old code that just looked at certain code point blocks was a lot more compact and independent of the Unicode version. I’m just not sure if it is standard conforming?  

Could we just use the old code combined with checking the category code to see if the code point is assigned?

stevengj replied:
(I still think we should use the data file, but only for tests.)

archermarx replied:
As i looked more into the emoji codepoints, i found some corner cases which weren't well-covered by the old version. It would be possible to patch those, but the secondary issue was that there's a bunch of unassigned emoji in the blocks in the old version. If we're ok with saying "yes this is an emoji" to characters in those larger blocks which are not (yet) emoji, then we can go with the old system. If not, we would either need to restrict those ranges to those codepoints which are currently assigned manually or via parsing the emoji_data file like i do here.

i do agree the old version was a lot simpler, so i'm not sure what the best solution is.

stevengj replied:
I agree that we should return `false` for currently unassigned codepoints, but we can check for that simply by returning `false` if `Unicode.category_code(char) == Unicode.UTF8PROC_CATEGORY_CN`

archermarx replied:
Oh neat, didn't know about that. I'll get back with a new version later today.

archermarx replied:
So looking at this, there are things in the emoji data file which fall under unicode category Sm and Po (◾ and ‼, respectively), and things in So which do not qualify as emoji (Ⓕ). To be honest, I'm not sure why some of these symbols get called emoji and some don't, but I think the only way to be complete here is to use the full list of ranges. We don't have to download it necessarily (could take the output from the file and manually include that in the unicode.jl file, but that's a 702-element array), but I'm not sure of another way to make sure we catch all emoji without false positives

archermarx committed:
Download when parsing emoji file

stevengj commented:
What is the `isemoji` function good for, especially if the set of "emoji" characters is a bit fuzzy and doesn't include all dingbats?

StefanKarpinski commented:
Another question: suppose we had a good definition of a single character being an emoji; given, what is the utility of `isemoji` on a string? Why would one want to know that a string only consists of emoji? Wouldn't it be more useful to ask if any of the characters in a string are emoji or not? The zero-width joiner issue complicates things, but it's a little unclear to me what the right predicate should be. Most questions you could ask about a string could presumably be answered by calling `graphemes` and then applying a predicate to individual graphemes. Maybe that's why this PRs predicate makes sense: it's only intended to be applied to single-grapheme strings?

archermarx commented:
After some thought and retooling yesterday, I agree. The utility most makes sense in identifying a single emoji, not whether a string is all emoji. I have a new version I'll upload in a bit that does that well and catches all the emoji in the dataset. We could separately have a `contains_emoji` function. My use case for this has been in parsing text and compiling statistics of emoji usage in large corpuses of internet conversation, so i think splitting out individual emoji makes far more sense than just saying "everything in this string is an emoji".

The remaining question is how do we want do the check for single characters? the most effective way is to parse emoji_data.txt, but do we want to download it? have a cached copy? a pre-made list of ranges in a separate file? i think the last would be most space efficient and would obviate the need for emoji file parsing utilities in the main body of the Unicode module and let them live in the tests instead

StefanKarpinski commented:
The problem with only having `isemoji` as a character predicate would seem to be that some emoji are not a single character but a whole grapheme, so if you want to count them you probably want to count such a sequence as a single emoji, right? One way to do this would be to call `graphemes` to iterate the graphemes in a string and then have a predicate that can be applied to each grapheme string to indicate whether it is an emoji grapheme or not. Of course, that's not the only way that predicate can be called, so the question is what is the most useful generalization from a predicate like that to arbitrary strings?

archermarx commented:
I have a version (to be uploaded shortly) that has isemoji as a character predicate _and_ a string predicate, so that we get this behavior

```julia
julia> teststring = "My family looks like this: 👩‍👩‍👦‍👦 and I ❤️ them"

julia> filter(isemoji, [g for g in graphemes(teststring)])
2-element Array{SubString{String},1}:
 "👩\u200d👩\u200d👦\u200d👦"
 "❤️"

archermarx committed:
Update isemoji function and add emoji_ranges file

archermarx commented:
Ok, I've uploaded my new version. Unrelated to the `isemoji` function, I found a bug in `graphemes`  which crashes julia

```julia
julia> using Unicode

julia> graphemes("👨🏻‍🤝‍👨🏽")
```
On my machine (Julia running in WSL on Windows 10) this causes the julia REPL to crash

archermarx commented:
The problem with the graphemes seems to be related to https://github.com/JuliaLang/julia/issues/37680, and stems from a bug in the utf8-proc library.

archermarx committed:
Remove incorrect emoji ranges

archermarx committed:
Support two-character keycap pattern

archermarx commented:
One last thing I'd want to add would be the following:

```julia
emojis(s::AbstractString) = Iterators.filter(isemoji, graphemes(s))
```

so we'd have

```julia
julia> collect(emojis("My family looks like this: 👩‍👩‍👦‍👦 and I ❤️ them"))
2-element Vector{SubString{String}}:
 "👩\u200d👩\u200d👦\u200d👦"
 "❤️"
```

archermarx committed:
Fix test, but one test in precompile is failing

StefanKarpinski commented:
I think the new behavior makes sense but now I worry about the name: "emoji" is both plural and singular, so it might be surprising to someone that this function considers a string that consists of two emoji to not be emoji. So perhaps this predicate should be called `isoneemoji`? Or perhaps it would be better to provide some kind of `find`-like interface instead, such as `findemoji(s::AbstratString, i::Integer)` which returns `nothing` if there are no emoji in the given string or a pair of indices if there are?

stevengj commented:
My basic concern here is about adding a nontrivial new function to the stdlib without clear idea of what it is for.  What is the application of this function?

StefanKarpinski commented:
I think the application is counting emoji in text. You would do it by calling `graphemes` and then calling `isemoji` on each grapheme. It seems like this might be better suited to a package which can evolve its API as necessary.

archermarx commented:
Yeah my main application for this has been parsing and analyzing message data to determine things like emoji frequencies, usage patterns etc. I have also been thinking a package might be easier for this at least for now.

vtjnash commented:
Closing, per apparent consensus among participants that this should be a separate package.
